### PR TITLE
[InstCombine] Reducing rust i128::midpoint instructions

### PR DESCRIPTION
**Pull request by wizardengineer**
**Issue**: https://github.com/rust-lang/rust/issues/124790

This seems to only affect `u128` data types

https://alive2.llvm.org/ce/z/QmFkhy
https://rust.godbolt.org/z/T7eKP3Tvo

### CODE REVIEW

**Comment by dtcxzyw** [comment]
Please read the guideline https://llvm.org/docs/InstCombineContributorGuide.html :)

File: llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll [hunk 1]
```diff
@@ -0,0 +1,152 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+define noundef i128 @xor_lshr_and(i128 noundef %x, i128 noundef %y) unnamed_addr #0 {
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll:4]:
```suggestion
define i128 @xor_lshr_and(i128 %x, i128 %y) {
```
Please remove unused attributes/metadata.

File: llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+; check-label: define noundef i128 @xor_lshr_and(
+; check-same: i128 noundef [[x:%.*]], i128 noundef [[y:%.*]]) unnamed_addr {
+; check-next:  [[start:.*:]]
+; check-next:    [[xor:%.*]] = xor i128 [[y]], [[x]]
+; check-next:    [[lshr:%.*]] = lshr i128 [[xor]], 1
+; check-next:    [[and:%.*]] = and i128 [[y]], [[x]]
+; check-next:    [[add:%.*]] = add i128 [[lshr]], [[and]]
+; check-next:    ret i128 [[add]]
+;
```
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll:13]:
```suggestion
```
It is not generated by `utils/update_test_checks.py`.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp [hunk 3]
```diff
@@ -1483,6 +1483,47 @@ static Instruction *foldBoxMultiply(BinaryOperator &I) {
...
+// with the i128 data type and reducing the amount of asm instructions
+// generated.
+static Instruction *foldMidpointExpression(BinaryOperator &I) {
+  Value *X, *Y;
+
+  if (!I.getType()->isIntegerTy(128))
+    return nullptr;
+
+  if (!match(&I, m_Add(m_LShr(m_Xor(m_Value(X), m_Value(Y)), m_ConstantInt()),
+                      m_And(m_Value(X), m_Value(Y)))))
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:1497]:
```suggestion
                      m_c_And(m_Deferred(X), m_Deferred(Y)))))
```

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll (hunk 1 above)
Comment (medievalghoul) [reply llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll]: fixed

File: llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll (hunk 2 above)
Comment (medievalghoul) [reply llvm/test/Transforms/InstCombine/xor_lshr_and_i128.ll]: done, fixed

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp (hunk 3 above)
Comment (medievalghoul) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: I made sure to fix it

Comment (medievalghoul) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: thank you!

**Comment by medievalghoul** [comment]
Hey @dtcxzyw, not to be a bother but I was wondering was there anything else i needed to change? :)

**Comment by dtcxzyw** [comment]
I am not sure whether it is appropriate to do this transform in InstCombine. It would be better to do this in SDAG.
cc @RKSimon @topperc

**Comment by medievalghoul** [comment]
> I am not sure whether it is appropriate to do this transform in InstCombine. It would be better to do this in SDAG.
> 
> 

Ah, I see. Just curious, why would I put this in SDAG rather than InstCombine?

**Comment by RKSimon** [comment]
This is just a avgflooru node, which DAG can already recognise and handle.
```
Optimized lowered selection DAG: %bb.0 'src:start'
SelectionDAG has 22 nodes:
  t0: ch,glue = EntryToken
    t20: i64 = extract_element t27, Constant:i64<0>
  t23: ch,glue = CopyToReg t0, Register:i64 $rax, t20
    t18: i64 = extract_element t27, Constant:i64<1>
  t25: ch,glue = CopyToReg t23, Register:i64 $rdx, t18, t23:1
      t6: i64,ch = CopyFromReg t0, Register:i64 %2
      t8: i64,ch = CopyFromReg t0, Register:i64 %3
    t10: i128 = build_pair t6, t8
      t2: i64,ch = CopyFromReg t0, Register:i64 %0
      t4: i64,ch = CopyFromReg t0, Register:i64 %1
    t9: i128 = build_pair t2, t4
  t27: i128 = avgflooru t10, t9
  t26: ch = X86ISD::RET_GLUE t25, TargetConstant:i32<0>, Register:i64 $rax, Register:i64 $rdx, t25:1
```
What we're missing is better legalization handling in DAGTypeLegalizer::ExpandIntRes_AVG - your code expansion should work pretty well in there instead of InstCombine, where we shouldn't be making assumptions about type legality.

**Comment by medievalghoul** [comment]
I understand, thank you for taking the time to explain it to me. I'll close this PR

**Comment by RKSimon** [comment]
@medievalghoul Are you happy to work on a DAG equivalent?

**Comment by medievalghoul** [comment]
@RKSimon I'd be more than happy to work on a DAG equivalent, I just started today.

**Pull request closed without merging**